File: gremlins/clients/protocol.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class UsageStats:
    prompt_tokens: int = 0
    completion_tokens: int = 0
    cached_input_tokens: int = 0
    cache_creation_input_tokens: int = 0
    reasoning_tokens: int = 0
    turns: int = 0


@dataclass
class CompletedRun:
    exit_code: int
    text_result: str | None = None
    events: list[dict[str, Any]] | None = None
    cost_usd: float | None = None
    token_usage: UsageStats | None = None

File: crates/gremlins/src/clients/protocol.rs
use serde::{Deserialize, Serialize};

/// Aggregated token usage summary from an agent run.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UsageStats {
    pub prompt_tokens: u64,
    pub completion_tokens: u64,
    pub cached_input_tokens: u64,
    pub cache_creation_input_tokens: u64,
    pub reasoning_tokens: u64,
    pub turns: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CompletedRun {
    pub exit_code: i32,
    pub text_result: Option<String>,
    pub events: Option<Vec<serde_json::Value>>,
    pub cost_usd: Option<f64>,
    pub token_usage: Option<UsageStats>,
}
